# Proyek Diabetes Prediction

* Nama: Alif Khusain Bilfaqih
* Username: akbf_apotheosis
* Email: faqihkeren44@gmail.com

## Data Loading

**Import Library Yang DIbutuhkan**

Untuk mengolah dan memvisualisasikan data hingga membuat model, dibutuhkan library yang perlu sisiapkan, seperti pandas, numpy, seaborn, matplotlib, dan sebagainya.

In [132]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
%matplotlib inline
import seaborn as sns

**Import dataset**

Sebelum kita memulai semuanya, tentunya kita harus memasukkan dataset yang ingin kita gunakan.
Alamat dataset yang digunakan sebelumnya sudah di upload di github, lalu diambil dengan memasukan nama 'url' nya.
Dataset yang kita gunakan ada 2 file, yaitu 'credit' dan 'movies'.

In [133]:
url_credits = "https://raw.githubusercontent.com/faqihkeren44/movie_recommendation/refs/heads/main/tmdb_5000_credits.csv"
url_movies = "https://raw.githubusercontent.com/faqihkeren44/movie_recommendation/refs/heads/main/tmdb_5000_movies.csv"

credits = pd.read_csv(url_credits)
movies = pd.read_csv(url_movies)

**Menggabungkan Dataset**

Agar lebih mudah mengolah data, kedua dataset kita gabungkan dengan memanfaatkan library pandas dengan menggunakan `.merge()`. Kedua dataset ini digabungkan sesuai kolom `title`.

Dataset sudah selesai kita masukan. Mari kita lihat isi 5 baris pertama dari dataset, dengan perintah `.head(3)`

In [134]:
df = pd.merge(credits, movies, on='title')
df.head(3)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   

                                                crew     budget  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...  237000000   
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  300000000   
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  245000000   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   

                                            keywords original_language  ...  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en  ...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en  ...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en  ...   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   

                                production_countries  release_date  \
0  [{"iso_3166_1": "US", "name": "United States o...    2009-12-10   
1  [{"iso_3166_1": "US", "name": "United States o...    2007-05-19   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...    2015-10-26   

      revenue runtime                                   spoken_languages  \
0  2787965087   162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...   
1   961000000   169.0           [{"iso_639_1": "en", "name": "English"}]   
2   880674609   148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...   

     status                                         tagline vote_average  \
0  Released                     Enter the World of Pandora.          7.2   
1  Released  At the end of the world, the adventure begins.          6.9   
2  Released                           A Plan No One Escapes          6.3   

  vote_count  
0      11800  
1       4500  
2       4466  

[3 rows x 23 columns]

Untuk melihat beberapa informasi feature pada data, bisa memanfaatkan kode `.info()` dan `describe()`.

Dapat dilihat bahwa setiap kolom memiliki 100000 data. Dari data tersebut ada yang berupa kategorikal, dan ada yang numerical.

In [135]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   movie_id              4809 non-null   int64  
 1   title                 4809 non-null   object 
 2   cast                  4809 non-null   object 
 3   crew                  4809 non-null   object 
 4   budget                4809 non-null   int64  
 5   genres                4809 non-null   object 
 6   homepage              1713 non-null   object 
 7   id                    4809 non-null   int64  
 8   keywords              4809 non-null   object 
 9   original_language     4809 non-null   object 
 10  original_title        4809 non-null   object 
 11  overview              4806 non-null   object 
 12  popularity            4809 non-null   float64
 13  production_companies  4809 non-null   object 
 14  production_countries  4809 non-null   object 
 15  release_date         

In [136]:
df.describe()

movie_id        budget             id   popularity       revenue  \
count    4809.000000  4.809000e+03    4809.000000  4809.000000  4.809000e+03   
mean    57120.571429  2.902780e+07   57120.571429    21.491664  8.227511e+07   
std     88653.369849  4.070473e+07   88653.369849    31.803366  1.628379e+08   
min         5.000000  0.000000e+00       5.000000     0.000000  0.000000e+00   
25%      9012.000000  7.800000e+05    9012.000000     4.667230  0.000000e+00   
50%     14624.000000  1.500000e+07   14624.000000    12.921594  1.917000e+07   
75%     58595.000000  4.000000e+07   58595.000000    28.350529  9.291317e+07   
max    459488.000000  3.800000e+08  459488.000000   875.581305  2.787965e+09   

           runtime  vote_average    vote_count  
count  4807.000000   4809.000000   4809.000000  
mean    106.882255      6.092514    690.331670  
std      22.602535      1.193989   1234.187111  
min       0.000000      0.000000      0.000000  
25%      94.000000      5.600000     54.000000  
50%     103.000000      6.200000    235.000000  
75%     118.000000      6.800000    737.000000  
max     338.000000     10.000000  13752.000000

## Data Preparation

**Menangani missing data**

Beberapa kesalahan yang ada pada data adalah missing value atau nilai yang tidak ada, dan data yang terduplikasi. Agar tidak menjadi masalah pada saat menggali informasi, atau visualisasi, atau yang lainnya, dapat dilihat dengan menggunakan kode `.isna().sum()` dan `.duplicated().sum()`

In [137]:
df.isnull().sum()

movie_id                   0
title                      0
cast                       0
crew                       0
budget                     0
genres                     0
homepage                3096
id                         0
keywords                   0
original_language          0
original_title             0
overview                   3
popularity                 0
production_companies       0
production_countries       0
release_date               1
revenue                    0
runtime                    2
spoken_languages           0
status                     0
tagline                  844
vote_average               0
vote_count                 0
dtype: int64

In [138]:
df.duplicated().sum()

0

Ternyata ada beberapa missing value, terutama pada kolom `homepage`. Namun kolom ini tidak akan kita gunakan, sehingga kita biarkan.

Data terduplikasi yang dicek dengan `.duplicated().sum()`, nilainya 0. Tapi sepertinya ada data terpulikasi yang tidak terhitung. Makanya kita cek judul yang sama di kolom `title`.

In [139]:
df[df['title'].duplicated()].sort_values(by='title')

movie_id            title  \
1361       268           Batman   
4272      2661           Batman   
4273      2661           Batman   
3651     39269  Out of the Blue   
3697     10844  Out of the Blue   
3698     10844  Out of the Blue   
973      72710         The Host   
2879      1255         The Host   
2880      1255         The Host   

                                                   cast  \
1361  [{"cast_id": 5, "character": "Jack Napier/The ...   
4272  [{"cast_id": 17, "character": "Batman / Bruce ...   
4273  [{"cast_id": 17, "character": "Batman / Bruce ...   
3651  [{"cast_id": 2, "character": "Don", "credit_id...   
3697  [{"cast_id": 12, "character": "Nick Harvey", "...   
3698  [{"cast_id": 12, "character": "Nick Harvey", "...   
973   [{"cast_id": 52, "character": "Melanie Stryder...   
2879  [{"cast_id": 3, "character": "Park Gang-du", "...   
2880  [{"cast_id": 3, "character": "Park Gang-du", "...   

                                                   crew    budget  \
1361  [{"credit_id": "52fe422fc3a36847f800aa4b", "de...   1377800   
4272  [{"credit_id": "52fe4363c3a36847f80509a7", "de...  35000000   
4273  [{"credit_id": "52fe4363c3a36847f80509a7", "de...   1377800   
3651  [{"credit_id": "52fe47099251416c9106826f", "de...         0   
3697  [{"credit_id": "52fe43c19251416c7501cceb", "de...         0   
3698  [{"credit_id": "52fe43c19251416c7501cceb", "de...         0   
973   [{"credit_id": "52fe487bc3a368484e0fa919", "de...  11000000   
2879  [{"credit_id": "52fe42eac3a36847f802ca6b", "de...  44000000   
2880  [{"credit_id": "52fe42eac3a36847f802ca6b", "de...  11000000   

                                                 genres  \
1361  [{"id": 10751, "name": "Family"}, {"id": 12, "...   
4272  [{"id": 14, "name": "Fantasy"}, {"id": 28, "na...   
4273  [{"id": 10751, "name": "Family"}, {"id": 12, "...   
3651                      [{"id": 18, "name": "Drama"}]   
3697                      [{"id": 18, "name": "Drama"}]   
3698                      [{"id": 18, "name": "Drama"}]   
973   [{"id": 27, "name": "Horror"}, {"id": 18, "nam...   
2879  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
2880  [{"id": 27, "name": "Horror"}, {"id": 18, "nam...   

                       homepage     id  \
1361                        NaN   2661   
4272                        NaN    268   
4273                        NaN   2661   
3651                        NaN  10844   
3697                        NaN  39269   
3698                        NaN  10844   
973   http://www.hostmovie.com/   1255   
2879                        NaN  72710   
2880  http://www.hostmovie.com/   1255   

                                               keywords original_language  \
1361  [{"id": 339, "name": "submarine"}, {"id": 849,...                en   
4272  [{"id": 848, "name": "double life"}, {"id": 84...                en   
4273  [{"id": 339, "name": "submarine"}, {"id": 849,...                en   
3651  [{"id": 2658, "name": "new zealand"}, {"id": 3...                en   
3697  [{"id": 4470, "name": "punk"}, {"id": 10183, "...                en   
3698  [{"id": 2658, "name": "new zealand"}, {"id": 3...                en   
973   [{"id": 1261, "name": "river"}, {"id": 1880, "...                ko   
2879  [{"id": 818, "name": "based on novel"}, {"id":...                en   
2880  [{"id": 1261, "name": "river"}, {"id": 1880, "...                ko   

      ...                               production_companies  \
1361  ...  [{"name": "Twentieth Century Fox Film Corporat...   
4272  ...  [{"name": "PolyGram Filmed Entertainment", "id...   
4273  ...  [{"name": "Twentieth Century Fox Film Corporat...   
3651  ...                                                 []   
3697  ...           [{"name": "Robson Street", "id": 71953}]   
3698  ...                                                 []   
973   ...  [{"name": "Cineclick Asia", "id": 685}, {"name...   
2879  ...  [{"name": "Nick Wechsler Productions", "id": 8...   
2880  ...  [{"

Ternyata benar, ada 3 judul yang sama dalam dataset kita. Mari kita persempit kolom dataset, dengan hanya melihat judul yang terduplikasi dan tanggal rilisnya.

In [140]:
# df[df.duplicated(subset=['title', 'release_date'], keep=False)].sort_values(by='title')
duplicated_titles = df[df['title'].duplicated()]
duplicated_titles[['title', 'release_date']].sort_values(by='title')

title release_date
1361           Batman   1966-07-30
4272           Batman   1989-06-23
4273           Batman   1966-07-30
3651  Out of the Blue   2006-10-12
3697  Out of the Blue   1980-05-01
3698  Out of the Blue   2006-10-12
973          The Host   2006-07-27
2879         The Host   2013-03-22
2880         The Host   2006-07-27

Dari tabel di atas, kita dapat menyimpulkan mungkin film tersebut dibuat ulang di tahun berikutnya dan ditayangkan dengan nama yang sama. Oleh sebab itu, dari ketiga film yang sama ini, kita hapus 1 saja, dengan perintah `.dop.duplitaces(subset=['title', 'release_date'])`.

In [141]:
df = df.drop_duplicates(subset=['title', 'release_date'])

In [142]:
drop_title = ['Batman', 'The Host', 'Out of the Blue']
duplicated_titles = df[df['title'].isin(drop_title)]
duplicated_titles[['title', 'release_date']]

title release_date
972          The Host   2013-03-22
973          The Host   2006-07-27
1360           Batman   1989-06-23
1361           Batman   1966-07-30
3650  Out of the Blue   1980-05-01
3651  Out of the Blue   2006-10-12

**Menambahkan Judul Rilis Film**

Ada film dengan judul yang sama, namun memiliki tanggal rilis yang berbeda (Film remake). Agar tidak mengurangi film yang direkomendasikan dan tidak membuat model bingung (karena ada 2 data yang sama), kita tambahkna tahun rilis film itu setelah nama film.

Langkah pertama yaitu mengubah kolom 'release_date' menjadi tanggal dengan `.to_datetime()`, lalu mengubah formatnya menjadi tahun saja.

In [143]:
df['release_date'] = pd.to_datetime(df['release_date'], errors='coerce').dt.year

<ipython-input-143-b958fd8b7b95>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['release_date'] = pd.to_datetime(df['release_date'], errors='coerce').dt.year


Selanjutnya, kita gabungkan judul dengan tahun rilisnya. Tahun rilis ini akan dimasukan kedalam tanda kurung (). Jika pada kolom 'release_date' tidak ada, maka hanya akan menampilkan nama film saja.

In [144]:
df['title'] = df.apply(
    lambda x: f"{x['title']} ({int(x['release_date'])})" if not pd.isna(x['release_date']) else x['title'],
    axis=1
)
df.head(3)

movie_id                                            title  \
0     19995                                    Avatar (2009)   
1       285  Pirates of the Caribbean: At World's End (2007)   
2    206647                                   Spectre (2015)   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   

                                                crew     budget  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...  237000000   
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  300000000   
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  245000000   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   

                                            keywords original_language  ...  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en  ...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en  ...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en  ...   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   

                                production_countries  release_date  \
0  [{"iso_3166_1": "US", "name": "United States o...        2009.0   
1  [{"iso_3166_1": "US", "name": "United States o...        2007.0   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...        2015.0   

      revenue runtime                                   spoken_languages  \
0  2787965087   162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...   
1   961000000   169.0           [{"iso_639_1": "en", "name": "English"}]   
2   880674609   148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...   

     status                                         tagline vote_average  \
0  Released                     Enter the World of Pandora.          7.2   
1  Released  At the end of the world, the adventure begins.          6.9   
2  Released                           A Plan No One Escapes          6.3   

  vote_count  
0      11800  
1       4500  
2       4466  

[3 rows x 23 columns]

**Menyiapkan Dataset Baru**

Karena ada begitu banyak kolom, kita membuat dataframe baru dengan nama `main_df`, yang berisi beberapa kolom saja, yaitu 'movie_id', 'title', 'cast', 'crew', 'genres', dan 'keywords'.

In [145]:
main_df = df[['movie_id', 'title', 'cast', 'crew', 'genres', 'keywords']].copy()
main_df.head()

movie_id                                            title  \
0     19995                                    Avatar (2009)   
1       285  Pirates of the Caribbean: At World's End (2007)   
2    206647                                   Spectre (2015)   
3     49026                     The Dark Knight Rises (2012)   
4     49529                               John Carter (2012)   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...   
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...   
2  [{"credit_id": "54805967c3a36829b5002c41", "de...   
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...   
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  
0  [{"id": 1463, "name": "culture clash"}, {"id":...  
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...  
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...  
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...  
4  [{"id": 818, "name": "based on novel"}, {"id":...

Kolom cast, crew, keywords, dan genres berisi dictionary yang tidak kesemuanya dibutuhkan. Oleh sebab itu, kita akan mengubah list dictionary pada beberapa kolom tersebut.

Hal pertama adalah mengubahnya menjadi list.

In [146]:
import ast

def convert (x):
  list = []
  for i in ast.literal_eval(x):
    list.append(i['name'])
  return list

Di atas sudah dibuatkan *function* untuk mengubah *dictionary* menjadi *list*, yang kita gunakan hanya untuk kolom 'genres' dan 'keywords'. Cara kerja *function* tersebut yaitu dengan mengambil kata dengan kata kunci 'name' dalam *dictionary*, dan memasukannya kedalam *list*. Hal itu terus berulang sampai semua kata dalam semua baris.

In [147]:
main_df['genres'] = main_df['genres'].apply(convert)
main_df['keywords'] = main_df['keywords'].apply(convert)
main_df.head()

movie_id                                            title  \
0     19995                                    Avatar (2009)   
1       285  Pirates of the Caribbean: At World's End (2007)   
2    206647                                   Spectre (2015)   
3     49026                     The Dark Knight Rises (2012)   
4     49529                               John Carter (2012)   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...   
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...   
2  [{"credit_id": "54805967c3a36829b5002c41", "de...   
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...   
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  
0  [culture clash, future, space war, space colon...  
1  [ocean, drug abuse, exotic island, east india ...  
2  [spy, based on novel, secret agent, sequel, mi...  
3  [dc comics, crime fighter, terrorist, secret i...  
4  [based on novel, mars, medallion, space travel...

Sebelum kita menindak kolom 'cast', mari kita lihat salah 1 isinya, yaitu pada index ke 3.

In [148]:
main_df['cast'].iloc[3]

'[{"cast_id": 2, "character": "Bruce Wayne / Batman", "credit_id": "52fe4781c3a36847f8139869", "gender": 2, "id": 3894, "name": "Christian Bale", "order": 0}, {"cast_id": 8, "character": "Alfred Pennyworth", "credit_id": "52fe4781c3a36847f8139881", "gender": 2, "id": 3895, "name": "Michael Caine", "order": 1}, {"cast_id": 5, "character": "James Gordon", "credit_id": "52fe4781c3a36847f8139875", "gender": 2, "id": 64, "name": "Gary Oldman", "order": 2}, {"cast_id": 3, "character": "Selina Kyle / Catwoman", "credit_id": "52fe4781c3a36847f813986d", "gender": 1, "id": 1813, "name": "Anne Hathaway", "order": 3}, {"cast_id": 4, "character": "Bane", "credit_id": "52fe4781c3a36847f8139871", "gender": 2, "id": 2524, "name": "Tom Hardy", "order": 4}, {"cast_id": 15, "character": "Miranda Tate", "credit_id": "52fe4781c3a36847f813988d", "gender": 1, "id": 8293, "name": "Marion Cotillard", "order": 5}, {"cast_id": 6, "character": "Blake", "credit_id": "52fe4781c3a36847f8139879", "gender": 2, "id": 2

Cast berisi nama karakter dalam film tersebut, dan aktor yang memerankannya, beserta id dan jenis kelamin. Biasanya, karakter penting seperti *main character* dan *main villain* berada di awal, kita ambil saja 5 nama karakter beserta aktor yang memerankannya.

Mari kita buat *function* untuk menjalankannya.

In [149]:
def actor_convert (x):
  list = []
  for i in ast.literal_eval(x):
    if len(list) < 5:
      list.append(i['name'])
    else:
      break
  return list

def character_convert (x):
  list = []
  for i in ast.literal_eval(x):
    if len(list) < 5:
      list.append(i['character'])
    else:
      break
  return list

Setelah dibuat *function*, kita masukan datanya, yaitu kolom 'cast', dengan perintah `.apply()`.

Karakter dalam film akan dimasukan kedalam kolom baru bernama 'character', dan pemerannya dimasukan kedalam kolom 'actor'. Jika sudah, maka kolom 'cast' sudah tidak diperlukan lagi, dan bisa kita buang.

In [150]:
main_df['actor'] = main_df['cast'].apply(actor_convert)
main_df['character'] = main_df['cast'].apply(character_convert)
main_df = main_df.drop('cast', axis=1)
main_df.head()

movie_id                                            title  \
0     19995                                    Avatar (2009)   
1       285  Pirates of the Caribbean: At World's End (2007)   
2    206647                                   Spectre (2015)   
3     49026                     The Dark Knight Rises (2012)   
4     49529                               John Carter (2012)   

                                                crew  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...   
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...   
2  [{"credit_id": "54805967c3a36829b5002c41", "de...   
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...   
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                               actor  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weave...   
1  [Johnny Depp, Orlando Bloom, Keira Knightley, ...   
2  [Daniel Craig, Christoph Waltz, Léa Seydoux, R...   
3  [Christian Bale, Michael Caine, Gary Oldman, A...   
4  [Taylor Kitsch, Lynn Collins, Samantha Morton,...   

                                           character  
0  [Jake Sully, Neytiri, Dr. Grace Augustine, Col...  
1  [Captain Jack Sparrow, Will Turner, Elizabeth ...  
2         [James Bond, Blofeld, Madeleine, M, Lucia]  
3  [Bruce Wayne / Batman, Alfred Pennyworth, Jame...  
4  [John Carter, Dejah Thoris, Sola, Tars Tarkas,...

Terakhir adalah kolom 'cast'. Mari kita lihat salah satu isinya pada index pertama dengan `.iloc[0]`.

In [151]:
main_df['crew'].iloc[0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

Banyak nama kru yang bertugas ada di dalam kolom 'crew', yang tidak semuanya penting. Karena tidak mungkin orang mencari film karena melihat siapa desain suaranya, kita hanya akan mengambil direkturnya saja.

In [152]:
def director_convert (x):
  list = []
  for i in ast.literal_eval(x):
    if i['job'] == 'Director':
      list.append(i['name'])
      break
  return list

Di atas adalah *function* untuk mengambil nama direktur dalam film. Kita jalankan dengan perintah `.apply()`, dan menghapus kolom 'crew'.

Berikut adalah hasil akhirnya

In [153]:
main_df['director'] = main_df['crew'].apply(director_convert)
main_df = main_df.drop('crew', axis=1)
main_df.head()

movie_id                                            title  \
0     19995                                    Avatar (2009)   
1       285  Pirates of the Caribbean: At World's End (2007)   
2    206647                                   Spectre (2015)   
3     49026                     The Dark Knight Rises (2012)   
4     49529                               John Carter (2012)   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                               actor  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weave...   
1  [Johnny Depp, Orlando Bloom, Keira Knightley, ...   
2  [Daniel Craig, Christoph Waltz, Léa Seydoux, R...   
3  [Christian Bale, Michael Caine, Gary Oldman, A...   
4  [Taylor Kitsch, Lynn Collins, Samantha Morton,...   

                                           character             director  
0  [Jake Sully, Neytiri, Dr. Grace Augustine, Col...      [James Cameron]  
1  [Captain Jack Sparrow, Will Turner, Elizabeth ...     [Gore Verbinski]  
2         [James Bond, Blofeld, Madeleine, M, Lucia]         [Sam Mendes]  
3  [Bruce Wayne / Batman, Alfred Pennyworth, Jame...  [Christopher Nolan]  
4  [John Carter, Dejah Thoris, Sola, Tars Tarkas,...     [Andrew Stanton]

Setelah semua menjadi *list*, selanjutnya mari kita ubah menjadi string biasa dengan memisahkan setiap isinya dengan koma (,).

In [154]:
features = ['genres', 'keywords', 'actor', 'character', 'director']

In [155]:
for col in features:
  main_df[col] = main_df[col].apply(lambda x: ', '.join(x) if isinstance(x, list) else x)
main_df.head(3)

movie_id                                            title  \
0     19995                                    Avatar (2009)   
1       285  Pirates of the Caribbean: At World's End (2007)   
2    206647                                   Spectre (2015)   

                                        genres  \
0  Action, Adventure, Fantasy, Science Fiction   
1                   Adventure, Fantasy, Action   
2                     Action, Adventure, Crime   

                                            keywords  \
0  culture clash, future, space war, space colony...   
1  ocean, drug abuse, exotic island, east india t...   
2  spy, based on novel, secret agent, sequel, mi6...   

                                               actor  \
0  Sam Worthington, Zoe Saldana, Sigourney Weaver...   
1  Johnny Depp, Orlando Bloom, Keira Knightley, S...   
2  Daniel Craig, Christoph Waltz, Léa Seydoux, Ra...   

                                           character        director  
0  Jake Sully, Neytiri, Dr. Grace Augustine, Col....   James Cameron  
1  Captain Jack Sparrow, Will Turner, Elizabeth S...  Gore Verbinski  
2           James Bond, Blofeld, Madeleine, M, Lucia      Sam Mendes

Setelah semuanya selesai, kita buat kolom baru dengan nama 'tags' ynag berisi dari gabungan semua kolom yang ada. Isi/nilai dri kolom 'tags' ini yang akan kita gunakan untuk menghitung rekomendasi yang tepat.

In [156]:
main_df['tags'] = main_df['genres'] + main_df['keywords'] + main_df['actor'] + main_df['character'] + main_df['director']
main_df['tags'].iloc[0]

'Action, Adventure, Fantasy, Science Fictionculture clash, future, space war, space colony, society, space travel, futuristic, romance, space, alien, tribe, alien planet, cgi, marine, soldier, battle, love affair, anti war, power relations, mind and soul, 3dSam Worthington, Zoe Saldana, Sigourney Weaver, Stephen Lang, Michelle RodriguezJake Sully, Neytiri, Dr. Grace Augustine, Col. Quaritch, Trudy ChaconJames Cameron'

Berikut adalah dataframe baru, yang bernama `new_df`, yang hanya berisi judul film (title), dan kolom 'tags'.

In [157]:
new_df = main_df[['title', 'tags']]
new_df.head(3)

title  \
0                                    Avatar (2009)   
1  Pirates of the Caribbean: At World's End (2007)   
2                                   Spectre (2015)   

                                                tags  
0  Action, Adventure, Fantasy, Science Fictioncul...  
1  Adventure, Fantasy, Actionocean, drug abuse, e...  
2  Action, Adventure, Crimespy, based on novel, s...

Agar tidak ada perbedaan, kita buat semua kata pada kolom 'tags' menjadi huruf kecil atau *undercase*.

In [158]:
new_df['tags'] = new_df['tags'].apply(lambda x: x.lower())
new_df['tags']

<ipython-input-158-1996f2866e26>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(lambda x: x.lower())


0       action, adventure, fantasy, science fictioncul...
1       adventure, fantasy, actionocean, drug abuse, e...
2       action, adventure, crimespy, based on novel, s...
3       action, crime, drama, thrillerdc comics, crime...
4       action, adventure, science fictionbased on nov...
                              ...                        
4804    action, crime, thrillerunited states–mexico ba...
4805    comedy, romanceedward burns, kerry bishé, mars...
4806    comedy, drama, romance, tv moviedate, love at ...
4807    daniel henney, eliza coupe, bill paxton, alan ...
4808    documentaryobsession, camcorder, crush, dream ...
Name: tags, Length: 4803, dtype: object

## Modeling and Result

Karena kita menggunakan TF-IDF, mari kita import library, lalu diinisialisasi dengan nama 'tf'.

In [159]:
from sklearn.feature_extraction.text import TfidfVectorizer
tf = TfidfVectorizer()

Hal pertama yang dilakukan adalah menghitung IDF pada kolom 'tags' dengan `tf.fit()`. Jika sudah, mari kita mapping array dari fitur index integer ke fitur utama dengan `tf.get_feature_names_out()`.

In [160]:
tf.fit(main_df['tags'])
tf.get_feature_names_out()

array(['10', '1000', '11', ..., '강연희', '이유진', '최민식'], dtype=object)

Setelah berhasil, kita *fit* kolom 'tags', dan mentranfsormasikannya kedalam bentuk matrix 2 dimensi. Kita bisa melihat ukurannya dengan perintah `.shape`.

In [161]:
tfidf_matrix = tf.fit_transform(new_df['tags'])
tfidf_matrix.shape

(4803, 38416)

Kita juga bisa melihat matriks tf-idf yang sudah kita buat dengan `tfidf_matrix.todense()`. Bisa juga kita buatkan dataframe agar lebih nyaman dilihatnya. Di bawah adalah contohnya, dengan sample acak (15 kolom dari 'tags', dan 10 baris nama film).

In [162]:
pd.DataFrame(
    tfidf_matrix.todense(),
    columns = tf.get_feature_names_out(),
    index = main_df.title
).sample(15, axis=1).sample(10, axis=0)

mcbrideseth  valarie  lackcharlie  \
title                                                                      
Superstar (1999)                               0.0      0.0          0.0   
J. Edgar (2011)                                0.0      0.0          0.0   
The Craft (1996)                               0.0      0.0          0.0   
Deuces Wild (2002)                             0.0      0.0          0.0   
Ace Ventura: When Nature Calls (1995)          0.0      0.0          0.0   
Bowling for Columbine (2002)                   0.0      0.0          0.0   
Micmacs (2009)                                 0.0      0.0          0.0   
To Kill a Mockingbird (1962)                   0.0      0.0          0.0   
The Boys from Brazil (1978)                    0.0      0.0          0.0   
Three to Tango (1999)                          0.0      0.0          0.0   

                                       valenti  breed  lian  cunningham  \
title                                                                     
Superstar (1999)                           0.0    0.0   0.0         0.0   
J. Edgar (2011)                            0.0    0.0   0.0         0.0   
The Craft (1996)                           0.0    0.0   0.0         0.0   
Deuces Wild (2002)                         0.0    0.0   0.0         0.0   
Ace Ventura: When Nature Calls (1995)      0.0    0.0   0.0         0.0   
Bowling for Columbine (2002)               0.0    0.0   0.0         0.0   
Micmacs (2009)                             0.0    0.0   0.0         0.0   
To Kill a Mockingbird (1962)               0.0    0.0   0.0         0.0   
The Boys from Brazil (1978)                0.0    0.0   0.0         0.0   
Three to Tango (1999)                      0.0    0.0   0.0         0.0   

                                       partyjoe  bergl   95  albin  \
title                                                                
Superstar (1999)                            0.0    0.0  0.0    0.0   
J. Edgar (2011)                             0.0    0.0  0.0    0.0   
The Craft (1996)                            0.0    0.0  0.0    0.0   
Deuces Wild (2002)                          0.0    0.0  0.0    0.0   
Ace Ventura: When Nature Calls (1995)       0.0    0.0  0.0    0.0   
Bowling for Columbine (2002)                0.0    0.0  0.0    0.0   
Micmacs (2009)                              0.0    0.0  0.0    0.0   
To Kill a Mockingbird (1962)                0.0    0.0  0.0    0.0   
The Boys from Brazil (1978)                 0.0    0.0  0.0    0.0   
Three to Tango (1999)                       0.0    0.0  0.0    0.0   

                                       smithchief  marjane  jerrygregory  \
title                                                                      
Superstar (1999)                              0.0      0.0           0.0   
J. Edgar (2011)                               0.0      0.0           0.0   
The Craft (1996)                              0.0      0.0           0.0   
Deuces Wild (2002)                            0.0      0.0           0.0   
Ace Ventura: When Nature Calls (1995)         0.0      0.0           0.0   
Bowling for Columbine (2002)                  0.0      0.0           0.0   
Micmacs (2009)                                0.0      0.0           0.0   
To Kill a Mockingbird (1962)                  0.0      0.0           0.0   
The Boys from Brazil (1978)                   0.0      0.0           0.0   
Three to Tango (1999)                         0.0      0.0           0.0   

                                       susann  
title                                          
Superstar (1999)                          0.0  
J. Edgar (2011)                           0.0  
The Craft (1996)                          0.0  
Deuces Wild (2002)                        0.0  
Ace Ventura: When Nature Calls (1995)     0.0  
Bowling for Columbine (2002)              0.0  
Micmacs (2009)                            0.0  
To Kill a Mockingbird (1962)           

### Cosine Similarity

Di atas, kita berhasil mengidentifikasi korelasi antara film dengan kolom 'tags'. Sekarang kita hitung derajat kesamaannya. Di sini, kita menggunakan `cosine_similarity` dari librarby `sklearn`.

In [163]:
from sklearn.metrics.pairwise import cosine_similarity
# from sklearn.metrics.pairwise import linear_kernel

In [164]:
cosine_sim = cosine_similarity(tfidf_matrix)
cosine_sim

array([[1.        , 0.01616338, 0.00855946, ..., 0.01567432, 0.        ,
        0.        ],
       [0.01616338, 1.        , 0.00522759, ..., 0.00807973, 0.01371672,
        0.        ],
       [0.00855946, 0.00522759, 1.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.01567432, 0.00807973, 0.        , ..., 1.        , 0.        ,
        0.        ],
       [0.        , 0.01371672, 0.        , ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

Untuk membuat struktur data yang memungkinkan kita mencari index dari sebuah judul film dengan cepat, kita buat data baru dengan nama `indices`.

In [165]:
indices = pd.Series(main_df.index, index = new_df['title']).drop_duplicates()

Selanjutnya, mari kita lihat matriks kesamaan setiap resto dengan menampilkan nama film dalam 5 sampel kolom (axis = 1) dan 10 sampel baris (axis=0). Jalankan kode berikut.

In [166]:
cosine_sim_df = pd.DataFrame(cosine_sim, index = new_df['title'], columns = new_df['title'])
cosine_sim_df.sample(5, axis = 1).sample(10, axis = 0)

title                                           Dude, Where’s My Car? (2000)  \
title                                                                          
Captain America: The First Avenger (2011)                           0.027517   
Tumbleweeds (1999)                                                  0.020135   
Can't Stop the Music (1980)                                         0.000000   
Miss Congeniality 2: Armed and Fabulous (2005)                      0.012692   
Small Soldiers (1998)                                               0.000000   
In the Valley of Elah (2007)                                        0.000000   
A Scanner Darkly (2006)                                             0.011096   
Firewall (2006)                                                     0.000000   
The Lost Boys (1987)                                                0.000000   
Joe Somebody (2001)                                                 0.000000   

title                                           Accidental Love (2015)  \
title                                                                    
Captain America: The First Avenger (2011)                     0.007470   
Tumbleweeds (1999)                                            0.006173   
Can't Stop the Music (1980)                                   0.003811   
Miss Congeniality 2: Armed and Fabulous (2005)                0.000000   
Small Soldiers (1998)                                         0.003465   
In the Valley of Elah (2007)                                  0.008746   
A Scanner Darkly (2006)                                       0.007930   
Firewall (2006)                                               0.010967   
The Lost Boys (1987)                                          0.000000   
Joe Somebody (2001)                                           0.000000   

title                                           The Monuments Men (2014)  \
title                                                                      
Captain America: The First Avenger (2011)                       0.078967   
Tumbleweeds (1999)                                              0.000000   
Can't Stop the Music (1980)                                     0.000000   
Miss Congeniality 2: Armed and Fabulous (2005)                  0.000000   
Small Soldiers (1998)                                           0.000000   
In the Valley of Elah (2007)                                    0.048411   
A Scanner Darkly (2006)                                         0.000000   
Firewall (2006)                                                 0.011951   
The Lost Boys (1987)                                            0.000000   
Joe Somebody (2001)                                             0.000000   

title                                           Gabriela (1983)  \
title                                                             
Captain America: The First Avenger (2011)              0.002712   
Tumbleweeds (1999)                                     0.000000   
Can't Stop the Music (1980)                            0.000000   
Miss Congeniality 2: Armed and Fabulous (2005)         0.000000   
Small Soldiers (1998)                                  0.000000   
In the Valley of Elah (2007)                           0.003175   
A Scanner Darkly (2006)                                0.000000   
Firewall (2006)                                        0.000000   
The Lost Boys (1987)                                   0.000000   
Joe Somebody (2001)                                    0.000000   

title                                           Universal Soldier: The Return (1999)  
title                                                                                 
Captain America: The First Avenger (2011)                                   0.009275  
Tumbleweeds (1999)                                                          0.000000  
Can't Stop the Music (1980)                                                 0.000000

### Mendapatkan Rekomendasi

Setelah kita memiliki data similarity (kesamaan) antar film, langkah terakhir adalah membuat fungsi agar sistem dapat menampilkan rekondasi film sesuai judul yang dimasukan.

In [167]:
def get_recommendations(title, cosine_sim=cosine_sim):
    # Mencari semua judul film yang cocok tanpa tahun
    matched_titles = main_df[main_df['title'].str.contains(title, case=False, na=False)]

    if matched_titles.empty:
        return f"Film dengan judul '{title}' tidak ditemukan dalam dataset."

    # Jika hanya ada satu film yang cocok
    if len(matched_titles) == 1:
        idx = matched_titles.index[0]
    else:
        # Menampilkan daftar film dengan tahun rilis untuk pemilihan
        print("Ditemukan beberapa judul film dengan nama yang sama:")
        print(matched_titles['title'])
        choice = input("Masukkan indeks film yang ingin direkomendasikan: ")

        try:
            idx = int(choice)
            if idx not in matched_titles.index:
                return "Indeks yang dimasukkan tidak valid."
        except ValueError:
            return "Harap masukkan indeks yang valid."

    # Mendapatkan skor kemiripan berdasarkan indeks
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:11]  # Mengambil 10 film teratas
    movie_indices = [i[0] for i in sim_scores]

    return main_df['title'].iloc[movie_indices]


Berikut penjelasannya:
- skor kemiripan film dimasukan kedalam list `sim_scores`.
- `enumerate`: Menambahkan index (0, 1, 2, ...) ke skor kemiripan sehingga setiap film memiliki pasangan.
- Mengurutkan daftar film berdasarkan kemiripan dalam urutan menurun (Skor tertinggi di atas)
- Hanya menampilkan 10 film dengan skor kemiripan tertinggi
- Membuat `movie_indices` yang berisi `i[0]`, yaitu hanya mengambil index dari pasangan (`index`, `similarity_score`)
- Terakhir, `return` atau mengembalikan judul film seuai *index* yang dihitung sebelumya.

Sekarang, mari kita coba menampilkan rekomendasi film Superman dan Avatar

In [168]:
get_recommendations('Batman')

Ditemukan beberapa judul film dengan nama yang sama:
9            Batman v Superman: Dawn of Justice (2016)
119                               Batman Begins (2005)
210                              Batman & Robin (1997)
299                              Batman Forever (1995)
428                              Batman Returns (1992)
1360                                     Batman (1989)
1361                                     Batman (1966)
3859    Batman: The Dark Knight Returns, Part 2 (2013)
Name: title, dtype: object
Masukkan indeks film yang ingin direkomendasikan: 1360


1361                                     Batman (1966)
210                              Batman & Robin (1997)
428                              Batman Returns (1992)
65                              The Dark Knight (2008)
119                               Batman Begins (2005)
3859    Batman: The Dark Knight Returns, Part 2 (2013)
3                         The Dark Knight Rises (2012)
9            Batman v Superman: Dawn of Justice (2016)
299                              Batman Forever (1995)
14                                 Man of Steel (2013)
Name: title, dtype: object

In [169]:
get_recommendations('Avatar')

838                      Alien³ (1992)
2405                     Aliens (1986)
3161                      Alien (1979)
278          Planet of the Apes (2001)
1533                  Moonraker (1979)
47      Star Trek Into Darkness (2013)
2200                    Lockout (2012)
2017                 Spaceballs (1987)
4338             Silent Running (1972)
305             Treasure Planet (2002)
Name: title, dtype: object

In [170]:
get_recommendations('Avengers')

Ditemukan beberapa judul film dengan nama yang sama:
7     Avengers: Age of Ultron (2015)
16               The Avengers (2012)
Name: title, dtype: object
Masukkan indeks film yang ingin direkomendasikan: 1


'Indeks yang dimasukkan tidak valid.'

## Evaluation

Evaluasi sistem rekomendasi berbasis *Content-Based Filtering* seperti yang kita buat dilakukan untuk mengukur seberapa baik sistem memberikan rekomendasi yang relevan bagi pengguna. Pengukuran metode ini berdasarkan *history* pribadi saya setelah menonton film 'Spider-Man' yang mungkin akan saya rekomendasikan juga pada orang lain.

In [171]:
def evaluate(recommended, actual):
    recommended_set = set(recommended)
    actual_set = set(actual)

    true_positives = len(recommended_set & actual_set)
    precision = true_positives / len(recommended_set) if recommended_set else 0
    recall = true_positives / len(actual_set) if actual_set else 0
    f1 = 2 * (precision * recall) / (precision + recall) if (precision + recall) > 0 else 0

    return precision, recall, f1

Di atas merupakan fungsi untuk mengukur metrik evaluasi *precision*, *recall*, dan *f1*. Daftar film hasil rekomendasi dan pengalaman pribadi saya diubah menjadi set, karena set memiliki kelebihan, yaitu hanya menyimpan elemen unik dan mendukung operasi himpunan, seperti union, intersection, dan difference, yang akan berguna untuk menghitung true positives.

- *precision* membagi jumlah yang film yang sama dengan total rekomendasi yang diberikan oleh sistem. Jika `recommended_set` kosong, maka bernilai 0.
- *recall* menghitung seberapa banyak item yang relevan yang ditemukan oleh sistem dari keseluruhan item yang relevan.
- *f1* menggabungkan precision dan recall menjadi satu metrik tunggal dengan membagi hasil perkalian *precision* dengan *recall*, dan penjumlahan *precision* dengan *recall*

Di bawah merupakan hasil setelah menjalankan fungsi `evaluate`, dengan membandingkan film hasil rekomendasi film 'Spider-Man' tahun 2002, dengan pengalaman saya pribadi.

In [172]:
recommended = get_recommendations('Spider-Man').tolist()
actual = ['Spider-Man 3 (2007)', 'The Amazing Spider-Man 2 (2014)',
          'Spider-Man 2 (2004)', 'The Amazing Spider-Man (2012)',
          'Avengers: Age of Ultron (2015)']
precision, recall, f1 = evaluate(recommended, actual)
print(f"Precision: {precision}, Recall: {recall}, F1-Score: {f1}")

Ditemukan beberapa judul film dengan nama yang sama:
5                  Spider-Man 3 (2007)
20       The Amazing Spider-Man (2012)
30                 Spider-Man 2 (2004)
38     The Amazing Spider-Man 2 (2014)
159                  Spider-Man (2002)
Name: title, dtype: object
Masukkan indeks film yang ingin direkomendasikan: 159
Precision: 0.4, Recall: 0.8, F1-Score: 0.5333333333333333


Dari hasil evaluasi di atas, sistem sudah cukup baik dalam menemukan sebagian besar item yang relevan (80% relevan ditemukan), namun hanya 40% yang relevan. Nilai f1-Score yang relatif lebih rendah (0.533) menunjukkan bahwa meskipun recall cukup baik, precision perlu ditingkatkan untuk mencapai keseimbangan yang lebih baik antara keduanya.

In [173]:
%timeit get_recommendations('Avatar')

9.63 ms ± 2.35 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


Terakhir adalah evaluasi waktu yang dibutuhkan sistem untuk menampilkan rekomendasi.

Di sini kita mencoba dengan membuat sistem menampilkan rekomendasi film 'Avatar' 10x, lalu menghitung jumlah rata-rata waktu yang dibutuhkan.

Dapat dilihat rata-rata waktu yang dibutuhkan untuk menjalankan 1 loop adalah 27 *milisecond*, dengan vasiari 3.3 *milisecond*.